Yocto devday update (#7)
3mdeb/conferences: yocto-update-strategies.md

]

---
# Who we are ?

.center[.image-15[![](remark-templates/3mdeb-presentation-template/images/coreboot-1024x1024.png)] .image-15[![](remark-templates/3mdeb-presentation-template/images/uefi-1024x1024.png)] .image-15[![](remark-templates/3mdeb-presentation-template/images/lvfs.png)] .image-15[![](remark-templates/3mdeb-presentation-template/images/yocto.png)]]
.center[.image-35[![](remark-templates/3mdeb-presentation-template/images/openpower.svg)]]

* coreboot licensed service providers since 2016 and leadership participants
* UEFI Adopters since 2018
* Yocto Participants and Embedded Linux experts since 2019
* Official consultants for Linux Foundation fwupd/LVFS project since 2020
* IBM OpenPOWER Foundation members since 2020

---
# Agenda

* Project lifecycle
* Project components
* Update strategy
* Performing update
  - build environment
  - external layers update
  - custom layer update
* Build troubleshooting
  - bootloader
  - kernel
  - default applications
  - custom applications
* Veryfing update
* Summary

---
# Project lifecycle

> * Initiation

Reviewer: It does not fit on the slide perfectly.
Also, we have lots of text in this presentation. This slide is perfect for making image out of it.

Author: added img instead of text https://github.com/3mdeb/conferences/pull/7/commits/c5880dcd8abc43cf78a29114a79cfb0bbd2354f8

project. This documentation helps prioritize updates based on project
requirements, ensuring that updates align with the desired goals of the system.

The build environment is an important component when it comes to updating
Yocto-based projects. An automated build environment is desirable as it improves
efficiency and consistency. The build environment should fulfill the
requirements specified in the Yocto Project documentation, ensuring
compatibility with the Yocto framework.

Tests are essential for validating the functionality of the project during
updates. It is crucial to have a comprehensive set of tests that cover all
custom functionalities within the project. Automating tests, if possible,
improves the frequency and reliability of testing, enabling faster and more
efficient updates.

Implementing CI/CD processes for build and tests is highly beneficial in
updating Yocto-based projects. CI/CD automation helps streamline the building
and testing of the project, reducing manual efforts and saving time. By
automating these processes, we ensure consistent and reliable results,
maintaining the integrity of the project throughout the update process.

---
# Project components

> .center[<img src="img/compo-life.svg" width="550px">]

Reviewer: No `.image` works well enough that you need to use HTML tag?

I'd use rounded rectangles for inner and outer, and add some color instead of white. It looks weird to me, but that may be just preference.

Also, the white background could be removed if possible, as the background on slides is not white. But that's minor.

Author: I forgot about .image TBH. I used your suggestions about rectangles and also removed background from used images. Change for this slide here https://github.com/3mdeb/conferences/pull/7/commits/89ebb501e60c3428e0697cd2add271b31a0c07ba

for efficient updates, ensuring compatibility with the latest Yocto releases.

During the testing phase, comprehensive tests are designed and implemented.
These tests validate the project's functionality and help us identify potential
regressions that may arise during updates. By conducting thorough tests, we can
ensure that the system remains stable and reliable after applying updates.

In the deployment phase, CI/CD processes come into play. These automated
workflows streamline the build and testing stages, making the deployment of
updates more efficient and reproducible. CI/CD ensures that updates are
integrated seamlessly into the system.

Finally, all of these project components, including architecture documentation,
features documentation, build environment, tests, and CI/CD, collectively
contribute to ongoing maintenance and updates. They form a cohesive framework
that supports the project throughout its lifecycle, enabling us to address
security vulnerabilities, minimize technical debt, and integrate new
functionalities as we maintain and update our Yocto-based projects.

---
# Update strategy

> * Remembering every project component during updates

Reviewer: Definitely too much text on this slide
Ideally if the bullet point is only one line long, maximum of two lines

we need some short phrases as bullet points, not necessarily the whole sentences

Author: I put here a little bit less text, fix https://github.com/3mdeb/conferences/pull/7/commits/47359d083f2e6c8919c2292127ced92fe98d1571

  - each project component, such as architecture documentation, features
    documentation, build environment, tests, and CI/CD, plays a crucial role in
    the project's functionality and stability
  - updates must consider all components to ensure compatibility, minimize
    regressions, and maintain the overall integrity of the Yocto-based project
* Importance of the update deployment method
  - the update deployment method impacts the efficiency, scalability, and user
    experience of the Yocto-based project
  - having a dedicated update system allows for streamlined and controlled
    updates, minimizing downtime and reducing the need for manual intervention
  - evaluating the requirements and constraints of the project helps determine
    whether a dedicated update system is necessary or if reflashing devices is
    feasible

???

TBD notes

---
# Update strategy

> * Yocto release cycle

Reviewer: Same as above
Wall of text on the slide is not the best way to go

Author: also fixed https://github.com/3mdeb/conferences/pull/7/commits/47359d083f2e6c8919c2292127ced92fe98d1571

 - updating based on the Yocto release cycle ensures access to the latest
   features, bug fixes, and security patches provided by the Yocto Project.
 - regularly synchronizing with the Yocto release cycle maintains a stable and
   up-to-date project, leveraging the community's collective efforts and
   expertise
* CVE checks and security reports
  - integrating CVE (Common Vulnerabilities and Exposures) checks into the Yocto
    build helps identify security vulnerabilities in the project's components
  - organizing updates based on security reports allows for timely patching and
    protection against known vulnerabilities
* Feature-set requirements
 - considering feature updates allows for tailored enhancements, ensuring the
   project meets evolving functional needs and remains competitive in the market

???

TBD notes
https://docs.yoctoproject.org/ref-manual/release-process.html

---
# Performing update

* Build environment

>   - ensuring the build environment is up-to-date and compatible with the Yocto

Reviewer: can split into two bullets

Author: done https://github.com/3mdeb/conferences/pull/7/commits/6c57cc87dfbde19cf65b4be94a4decda28bde8fc

    Project requirements
* External layers update
  - updating external layers used in the project to their latest versions
* Custom layer update

>   - updating the custom layer(s) specific to the project, incorporating

Reviewer: possibly can split into two bulelts as well

Author: done https://github.com/3mdeb/conferences/pull/7/commits/6c57cc87dfbde19cf65b4be94a4decda28bde8fc

    necessary changes and improvements
* Real-life examples
  - responsive multimedia casting device using RaspberryPi
  - Yocto BSP maintenance
  - secure gateway for IoT data transfer using i.MX6

>   - full build automation in project using Rockchip RK3399

Reviewer: This slides looks much better in general, than two previous ones, in terms of text capacity

Author: Roger that

smooth update process.

Next, we need to update the external layers used in the project. External layers
provide additional functionality and components to the Yocto-based project. It's
important to check for updates to these external layers and incorporate the
latest versions to benefit from bug fixes, security patches, and new features.

Finally, we have the custom layer update. Custom layers are specific to the
project and may contain project-specific modifications, configurations, or
recipes. During the update process, we need to review and update these custom
layers to incorporate necessary changes, improvements, or bug fixes. This
ensures that the project's customizations remain compatible and functional with
the updated system.

By addressing these three aspects—build environment, external layers, and custom
layer update—we can ensure a smooth and successful update process for
Yocto-based projects. Updating these components in a systematic manner reduces
the risk of compatibility issues, ensures the project remains up to date, and
allows for seamless integration of new functionalities and improvements.

---
# Performing update - build environment

> * Verify the compatibility of the build environment with the desired Yocto

Reviewer: Could fit into one line if shortened

Author: whole slide replaced with image https://github.com/3mdeb/conferences/pull/7/commits/5d224bb9561a6873714203312f58eea470e81930

  Project release
  - see `Build Host Packages` in Mega Manual
  - use correct release
* Prepare reproducible build environment
  - ansible can be used to put together such an environment
  - other possibility is to prepare containeraized image; e.g.
    [kas](https://github.com/siemens/kas),
    [imx-docker](https://github.com/nxp-imx/imx-docker),
    [yocto-docker](https://github.com/3mdeb/yocto-docker)
  - easier to distribute such an environment among colleagues

> * Review and update the configuration files (local.conf, bblayers.conf) to

Reviewer: ```suggestion
* Review and update the configuration files (`local.conf`, `bblayers.conf`) to
```

>   reflect any changes or additions required for the update

Reviewer: Maybe this can be bullet point under the `Review ...`?

Author: whole slide replaced with image https://github.com/3mdeb/conferences/pull/7/commits/5d224bb9561a6873714203312f58eea470e81930

  - see release notes, to check if any changes need to be applied
  - rethink the way build configuration files are stored

???

rethink the way build configuration files are stored - you may want to go back
to the version before an update

---
# Performing update - external layers update

.center[<img src="img/project-layers.svg" width="450px">]

* Project consist of external and custom layers

>   - this way we follow the rule **Never modify the POKY layer. Never. Ever.**

Reviewer: ```suggestion
  - this way we follow the rule **Never modify the POKY layer. Never.**
```

Author: Removed this and left only an image https://github.com/3mdeb/conferences/pull/7/commits/85813329f1e5bcdea81c19b1f8d9f88cda102040#diff-a1666090e402e433c955cc46797dc399867480b4d815574efd9f5d4181e24ac3R319

  - any customizations goes into `meta-custom_layer`: we have full control on
    what and why was changed through `.bbappends`
* For `external` we should pick Yocto release; specific commit hash

>   - task for that should be short, e.g. when using NXP BSP as base, we should

Reviewer: `task for that should be short` - I do not get what you meant here.

Author: now it should be clearer https://github.com/3mdeb/conferences/pull/7/commits/85813329f1e5bcdea81c19b1f8d9f88cda102040#diff-a1666090e402e433c955cc46797dc399867480b4d815574efd9f5d4181e24ac3R328

    follow hashes used in given BSP release

???

Never modify the POKY layer. Never. Ever. - even expand it on other layers

---
# Performing update - custom layer update

> * Most of the effort of performing updates should be implemented in custom

Reviewer: ```suggestion
* Most of the effort should be spent in custom
```

Reviewer: Or shorten in other way, to fit on the slide

  layers
* We can list some of the basic tasks carried out when updating Yocto, which
  should be done in custom layers
  - updating `LAYERSERIES_COMPAT` variable
  - going through the bbappends files and verifying that they all still apply
    (whether they depended on a revised version of a recipe)
  - verification of custom patches for used packages, perhaps some of them were
    added in the upstream
  - syntax and variables verification
  - path verification for the keywords `require` and `include`
* Check the migration guides in case we encounter unexpected behavior
  - there is one per every release
  - we may found there usefull scripts, like one for override syntax change

---
# Build troubleshooting

* We can split up every operating system into couple pieces

> .center[<img src="img/boot-process.svg" width="400px">]

Reviewer: This image could be bigger

Author: it is now https://github.com/3mdeb/conferences/pull/7/commits/f7e23ce4cb7a83fd0804e1f30ca618d34e5e63d2#diff-a1666090e402e433c955cc46797dc399867480b4d815574efd9f5d4181e24ac3R370

Author: aaand now the background is also removed https://github.com/3mdeb/conferences/pull/7/commits/0cc11132e21093e7af972bb0901b16bfb0c8f446

> * Updating whole system, means that every part needs to work correctly (meaning,

Reviewer: I would move these to speaker notes, so image can be bigger, and we can focus on your speech, not on reading slide.

You can leave some short bulleets if needed.

Author: bullets moved to notes https://github.com/3mdeb/conferences/pull/7/commits/f7e23ce4cb7a83fd0804e1f30ca618d34e5e63d2#diff-a1666090e402e433c955cc46797dc399867480b4d815574efd9f5d4181e24ac3R374

  as expected, presented in requirements)
  - bootloader - configuration, environment variables, boot script,
    functionality, drivers, boot-time
  - kernel/device-tree - hardware description, configuration, drivers, boot-time
  - custom/default apps - configuration, compilation, installation
  - functionalities - testing, feedback, improvements
* Other than that, we may also need to update our tests suites and CI/CD
  pipelines

---
# Build troubleshooting

> * In the following slides, we want to present some of the problems we have had

Reviewer: Try to shorten these

Author: done https://github.com/3mdeb/conferences/pull/7/commits/f231f9bc5faadeb08e6cb4a6ef9f40604f21a370

  to work on in the recent past
  - most of them consist of updating to Yocto Kirkstone release
  - our experience allows us to describe a wide range of problems that any Yocto
    developer may encounter when updating a custom system
* We will discuss the following problems
  - changes in bootloaders and kernels on i.MX based platforms
  - proper configuration of default apps provided to our systems
  - modifications to the source code of custom applications when updating
    systems that use Qt
  - fixing CI/CD pipelines after switching to new Yocto release
  - going through test suites to preserve functionalities checking

---
# Build troubleshooting - bootloader

> * Problems with updating bootloader may consist of

Reviewer: ```suggestion
* Example problems with updating bootloader

>   - correcting bootscript

Reviewer: Is this problem, or solution?

Author: yeah, smth not right here, now should be better https://github.com/3mdeb/conferences/pull/7/commits/0eab34f24aebae4de8e465a7c63e0c3d8bc543f4#diff-a1666090e402e433c955cc46797dc399867480b4d815574efd9f5d4181e24ac3R421

  - properly rebasing custom drivers and defconfigs

>   - Yocto recipe compilation, but should not be a big issue as long as we are

Reviewer: conisder dropping `but should not be a big issue as long as we are
    able to compile it manually`

Author: moved to notes https://github.com/3mdeb/conferences/pull/7/commits/0eab34f24aebae4de8e465a7c63e0c3d8bc543f4#diff-a1666090e402e433c955cc46797dc399867480b4d815574efd9f5d4181e24ac3R441

+			  <&clk IMX8MM_CLK_PCIE1_PHY>,
+			  <&clk IMX8MM_CLK_PCIE1_CTRL>;
+	assigned-clock-rates = <10000000>, <100000000>, <250000000>;
+	assigned-clock-parents = <&clk IMX8MM_SYS_PLL2_50M>,
+				 <&clk IMX8MM_SYS_PLL2_100M>,
+				 <&clk IMX8MM_SYS_PLL2_250M>;
```]

* device-tree II

.code-14px[```bash
&sai2 {
        pinctrl-names = "default";
        pinctrl-0 = <&pinctrl_sai2>;
        pinctrl-1 = <&pinctrl_codec_en>;
-       clocks = <&clk IMX8MM_CLK_SAI2_ROOT>;
        assigned-clocks = <&clk IMX8MM_CLK_SAI2>;
        assigned-clock-parents = <&clk IMX8MM_AUDIO_PLL1_OUT>;
        assigned-clock-rates = <24576000>;
        (...)
```]

???

Add `assigned-clocks`, `assigned-clock-rates` and
`assigned-clock-parents` nodes to `pcie0`.

mimics
https://gitlab.com/3mdeb/dave_eu/linux-imx/-/commit/0d5920e4c35d15a9999ee19bb0521f21158072a0
Specify the clock rate and parent of i.MX8MQ/MM PCIe clocks.

sai2 overwriting correctly setting clocks in other file which caused whole
platform hang while booting

---
# Build troubleshooting - default applications

* Default applications as a part of created image
  - by default we mean here part of systems to which we did not provide custom,

>     Client's source code

Reviewer: ```suggestion
    source code
```

  - issues here: configuration changes, compilation errors

* Configuration changes
  - missed or changed PACKAGECONFIG
  - custom `do_configure` functions

.code-14px[```bash
- PACKAGECONFIG += "ssl libssh2"
+ PACKAGECONFIG += "openssl libssh2"
```]

> tensorflow https://github.com/nxp-imx/meta-imx/pull/16

Reviewer: Bullet missing? This issue also needs a reply from us

Author: moved to footnote, I want to reproduce that only with BSP release, will do before DevDay

https://github.com/3mdeb/conferences/pull/7/commits/e94ac909b1bc4d17c54aec915e9ef92e099235be#diff-a1666090e402e433c955cc46797dc399867480b4d815574efd9f5d4181e24ac3R563

---
# Build troubleshooting - default applications

* Compilation errors

.code-14px[```bash
/tmp/(...)/python/interpreter_wrapper/interpreter_wrapper_pybind11.cc
| /work/build-wayland/tmp/(...)/python/interpreter_wrapper/interpreter_wrapper_pybind11.cc:16:10:
    fatal error: pybind11/functional.h: No such file or
directory
| 16 | #include "pybind11/functional.h"
(...)
| In file included from /work/build-wayland/tmp/(...)/interpreter_wrapper.cc:15:
| /work/build-wayland/tmp/(...)/python/interpreter_wrapper/interpreter_wrapper.h:28:10:
    fatal error: Python.h: No such file or directory

> | 28 | #include <Python.h>

Reviewer: Can we show how it can be fixed? Or at least, where to look for solution?

| | ^~~~~~~~~~
| compilation terminated.
| ninja: build stopped: subcommand failed.
```]

---
# Build troubleshooting - custom applications

> * Custom applications often are provided by our Clients

Reviewer: ```suggestion
* Custom applications often are provided by our clients
```

  - they were compiled in some environment
  - we integrate them for given Yocto release

>   - update may introduce problems (compilator flags, deprecated variables)

Reviewer: shorten so it fits in one line. This way the code will better fit in the slide as well

Author: done https://github.com/3mdeb/conferences/pull/7/commits/471facce24fae0adab35f05332fba3bfa4009929

  - example: using app after Qt5 update to Qt6

> .code-9px[```bash

Reviewer: Do we have 9px style in presentation template? The code looks really big, not as 9px in the slide.

Author: thanks for pointing that out, we have 9px in remark-templates but in other dirs, we could probably merge those templates one day..

now I corrected styles in whole presentation (was using 14px but there is no such thing https://github.com/3mdeb/conferences/pull/7/commits/73c0761e2646b80f04f58444b762938d65bebd71

| /build/tmp-glibc/(...)/git/libs/qmqtt_client/ecc/ecc_manager.cpp: At global scope:
| /build/tmp-glibc/(...)/git/libs/qmqtt_client/ecc/ecc_manager.cpp:42:13: warning: 'bool
    writePrivateKeyToFile(const QString&, EVP_PKEY*)' defined but not used [-Wunused-function]
|    42 | static bool writePrivateKeyToFile(const QString& fileName, EVP_PKEY* key)
|       |             ^~~~~~~~~~~~~~~~~~~~~
| make[2]: *** [Makefile:318: temp/ecc_manager.o] Error 1
| make[2]: Leaving directory '/build/tmp-glibc/(...)/build/libs/qmqtt_client'
| make[1]: *** [Makefile:47: sub-qmqtt_client-make_first-ordered] Error 2
| make[1]: Leaving directory '/build/tmp-glibc/(...)/build/libs'
| make: *** [Makefile:48: sub-libs-make_first-ordered] Error 2
| ERROR: oe_runmake failed
| WARNING: exit code 1 from a shell command.
ERROR: Task (/work/meta-abcd-app/recipes-qt/abcd-mqtt-proxy/abcd-mqtt-proxy_git.bb:do_compile)
    failed with exit code '1'
NOTE: Tasks Summary: Attempted 4630 tasks of which 4609 didn't need to be rerun and 1 failed.

Summary: 1 task failed:
  /work/meta-abcd-app/recipes-qt/abcd-mqtt-proxy/abcd-mqtt-proxy_git.bb:do_compile
```]

---

> # Veryfing update

Reviewer: Update Verification would fit better. Also in previous instances.

* Solving any build problems usually leads us to a bootable image
  - it is insufficient, many functionalities may not work as they should
* Documentation and testing are key here
  - docs allow us to understand expecting behavior
  - test should cover every funtionality and verify them
  - they also should fullfill agreeded requirements
* This is where the most tedious work of updating begins
  - iterations of tests, patches, rebuilds and retests
  - CI/CD pipeline helps a lot
* Tests suites needs also updates
  - the test result based on an erroneous assumption, which was changed with the
    update
  - logs changed behavior

---
# Veryfing update

* Examples of changes needed in tests, when we got different logs

.code-14px[```bash
-   path: /usr/bin/tensorflow-lite-2.4.0/examples
+   path: /usr/bin/tensorflow-lite-2.9.1/examples
```]

.code-14px[```bash
-   - "PCI bridge: Synopsys, Inc. DWC_usb3 (rev 01)"
+   - "PCI bridge: Synopsys, Inc. DWC_usb3 / PCIe bridge (rev 01)"
```]

* Change the way how output is achieved
  - logging level may changed so, we need to run `dmesg`
  - other commands may not longer print anything to userspace

---
# Summary

> * Performing Yocto metadata updates is a complex and multi-level problem

Reviewer: Try to shorten text in the slide, please

Author: done https://github.com/3mdeb/conferences/pull/7/commits/ee8eb3bdd8a1804b329190b6310ff5dcf615c84b
